dashboard/peers.go

```diff
@@ -0,0 +1,617 @@
+// them by the node ID. The peers failing before the handshake (unknown peers)
+// only have IP addresses, so their connection attempts are stored as part of the
+// value of the outer map.
+//
+// Another criteria is to limit the number of metered peers so that
+// they don't fill the memory. The selection order is based on the
+// peers activity: the peers that are inactive for the longest time
+// are thrown first. For the selection a fifo list is used which is
+// linked to the bottom of the peer tree in a way that every activity
+// of the peer pushes the peer to the end of the list, so the inactive
+// ones come to the front. When a peer has some activity, it is removed
+// from and reinserted into the list. When the length of the list reaches
+// the limit, the first element is removed from the list, as well as from
+// the tree.
+//
+// The active peers that are still connected have priority over the disconnected
+// ones, therefore the list is extended by a separator, which is a pointer to a
+// list element. The separator separates the active peers from the inactive ones,
+// and it is the entry for the list. If the peer that is to be inserted is active,
+// it goes after the separator, otherwise it goes before. This way the active peers
+// never move to the front before the inactive ones.
+//
+// The peers that don't manage to make handshake are not inserted into the list,
+// only their connection attempts are appended to the array belonging to their IP.
+// In order to keep the fifo principle, a super array contains the order of the
+// attempts, and when the overall count reaches the limit, the earliest attempt is
+// removed from the beginning of its array.
+//
+// This data structure makes it possible to marshal the peer
+// history simply by passing it to the JSON marshaler.
+type PeerContainer struct {
+	// Bundles is the outer map using the peer's IP address as key.
+	Bundles map[string]*PeerBundle `json:"bundles,omitempty"`
+
+	// activeSeparator is a pointer to the last inactive peer element, splitting
+	// the list into an inactive and an active part, and forming the entry for
+	// the peer list.
+	activeSeparator *list.Element
+
+	// knownPeers contains the peers that managed to make handshake.
+	knownPeers *list.List
+
+	// unknownPeers is the super array containing the IP addresses, from which
+	// the peers attempted to connect then failed before/during the handshake.
+	// Its values are appended in chronological order, which means that the
+	// oldest attempt is at the beginning of the array. When the first element
+	// is removed, the first element of the related bundle's attempt array is
+	// removed too, ensuring that always the latest attempts are stored.
+	unknownPeers []string
+
+	// geodb is the geoip database used to retrieve the peers' geographical location.
+	geodb *GeoDB
+
+	// refresh is the refresh rate used to generate the
+	// initial auxiliary traffic samples' time stamps.
+	refresh time.Duration
```

Comment:
This doesn't really make sense in the current incarnation of the dashboard any more. Lets get rid of the chart timestamps instead of complicating code further.